### PR TITLE
[compound][mgx-68] add compound test scenario

### PR DESCRIPTION
**Pull request by vanderian**

### CODE REVIEW

File: utils/eventListeners.ts [hunk 1]
```diff
@@ -127,3 +129,88 @@ export function findEventData(
     .filter((event) => `${event.section}.${event.method}` === method)
     .map((event) => event.event.toHuman().data);
 }
+
+export const awaitEvent = async (
```
Comment (goncer) [inline utils/eventListeners.ts:133]: plz rename it to : `waitForEvent` , will be easier to find it I think.

File: utils/eventListeners.ts [hunk 2]
```diff
@@ -127,3 +129,88 @@ export function findEventData(
...
+        }
+        if (counter === blocks) {
+          reject(`method ${method} not found within blocks limit`);
+        }
+      });
+    });
+  });
+};
+
+export const signSendSuccess = async (
```
Comment (goncer) [inline utils/eventListeners.ts:169]: why not use SignTx from sdk? I think that this could lead to incorrect results because of shuffling, right @mateuszaaa ?

**Review by goncer — CHANGES_REQUESTED** [review]
Hi @vanderian  I added a few comments, plz review them. 
If this script is not aiming to be included in the pipelines, plz fix only the util class comments and move this script to  `testPad` folder, that is where all manual-test scripts are.

File: test/story/story.compound.test.ts [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+                },
+              ],
+            },
+          },
+        },
+        4_000_000_000
+      ),
+      alice
+    );
+  });
```
[the remark this answers is not in the document]
Comment (vanderian) [reply test/story/story.compound.test.ts:116]: just a setup, not a test per se
will move it to beforeAll and add a comment

File: test/story/story.compound.test.ts [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+      await oakApi.api.rpc.automationTime.getTimeAutomationFees("XCMP", 1);
+    testLog
+      .getLog()
+      .info(`auto-compound: automationTime fee = ${automationTimeFee}`);
+
+    // xcm transact fee in TUR on MG
+  });
+});
+
+class OakApi {
```
[the remark this answers is not in the document]
Comment (vanderian) [reply test/story/story.compound.test.ts:334]: y

File: utils/Framework/Node/OakNode.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      feePerSecond: 537_600_000_000,
+      instructionWeight: 150_000_000 * 6,
+    });
+  }
+
+  constructor(api: ApiPromise) {
+    this.api = api;
+  }
+
+  static async create(uri: string): Promise<OakApi> {
```
Comment (chrisli30) [inline utils/Framework/Node/OakNode.ts:19]: You shouldn’t need to create those definitions yourselt. Importing oak.js should do.
```
import '@oak-foundation/api-augment'
import { rpc } from '@oak-foundation/types';
import { ApiPromise, WsProvider } from '@polkadot/api'
```


Please check out this quick demo, https://docs.oak.tech/docs/oak-js/
Oak.js Code, https://github.com/OAK-Foundation/oak.js

File: testPad/compound.testpad.test.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    const taskId = "compound";
+    // const seconds = Math.trunc(new Date().getTime() / 1000);
+    // const nextHour = seconds - (seconds % HOUR) + HOUR;
+
+    const executions = [0];
+    const txCreate = oakApi.api.tx.automationTime.scheduleXcmpTask(
+      // task id
+      oakApi.api.createType("Vec<u8>", taskId),
+      // intervals for task execution, 0 == immediate execution, only with 'features=dev_queue' oak build
+      { Fixed: { executionTimes: executions } },
```
Comment (chrisli30) [inline testPad/compound.testpad.test.ts:161]: @vanderian we are defining a new interface to call here, but the interface will be very similar to this one.
https://github.com/OAK-Foundation/oak.js/blob/main/demo/src/scheduleFixedDynamicDispatchTask.js#L67

```
const extrinsic = api.tx.automationTime.scheduleDynamicDispatchTask(providedId, { fixed: { executionTimes }}, call);
```

The `call` here is the function to be trigger on your blockchain by XCM. I picture the call to be something like `ClaimSwapDeposit`. Please check out my documentation of "TODOs for Mangata"
https://docs.google.com/document/d/1oGuHeVT9z6xwCqGwioL9P515T8wz9nsFzMmd8CWqXBE/edit#heading=h.6mqvvr281phu

Basically, in order to auto-compound, you need to create a new extrinsic to bind the Claim, Swap and Deposit in one call, and calculate how much to swap for max efficiency.

and then, in this line that call will be passed to `automationTime.scheduleXCMDynamicDispatchTask` to be inserted into Turing Network’s task registry, and we will trigger and send a message to you at the right conditions.

I’m also attaching the proposed flow chart here. Please feel free to comment on the above Google Docs.
Turing __ Mangata XCM Flow - Page 1

File: testPad/compound.testpad.test.ts [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+      .mul(new BN(currencyData.unwrap().feePerSecond))
+      .div(weightInSeconds);
+
+    // const xcmExecutionFee = new BN(4_000_000_000) // TUR base unit cost weight * 4 instructions in transfer
+    //   .mul(new BN(419000000000)) // TUR fps
+    //   .div(weightInSeconds);
+
+    const automationTimeFee =
+      // @ts-ignore
+      await oakApi.api.rpc.automationTime.getTimeAutomationFees("XCMP", 1);
```
Comment (chrisli30) [inline testPad/compound.testpad.test.ts:190]: The parameter of this call is incomplete yet. We will create a new call and demo code for you to return the correct fee estimate, but we need to decide the function to be triggerd by XCM, aka the `ClaimSwapDeposit`. Please refer to the above comment.

**Review by chrisli30 — COMMENTED** [review]
The new way to call OAK for scheduling an automation task is in this demo.
https://github.com/OAK-Foundation/oak.js/blob/main/demo/src/scheduleFixedDynamicDispatchTask.js#L67

Please note that in my current design, it’s the application(MangataX DEX) which would call our scheduling function, but not Mangata chain.

We try to off-load as much execution as possible to the application layer as XCM cost between chains are heavy.

Let us know what you think.

File: utils/Framework/Node/OakNode.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import { ApiPromise, WsProvider } from "@polkadot/api";
+import { Extrinsic } from "../../setup";
+
+export class OakApi {
```
Comment (goncer) [inline utils/Framework/Node/OakNode.ts:4]: why the class has different name than the file name?

File: testPad/compound.testpad.test.ts [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+      alice
+    );
+
+    await signSendSuccess(oakApi.api, txCreate, user1);
+
+    testLog.getLog().info("running section: await execute task");
+    // check the xcm automation success on oak
+    await waitForEvent(oakApi.api, "automationTime.XcmpTaskSucceeded");
+    // check that we have minted some liquidity
+    await waitForEvent(api, "xyk.LiquidityMinted");
```
[the remark this answers is not in the document]
Comment (goncer) [reply testPad/compound.testpad.test.ts:247]: nvm, realized that this is a testPad file. am I understanding right that this is only for manual checks? If thats true, feel free to resolve all the conversations on this specific file.

**Review by goncer — CHANGES_REQUESTED** [review]

File: testPad/compound.testpad.test.ts [hunk 10]
```diff
@@ -0,0 +1,329 @@
+import { getEnvironmentRequiredVars } from "../utils/utils";
+import { User } from "../utils/User";
+import { alice, api, setupApi, setupUsers, sudo } from "../utils/setup";
+import { testLog } from "../utils/Logger";
+import {
+  waitForEvent,
+  signSendFinalized,
+  signSendSuccess,
+} from "../utils/eventListeners";
```
[the remark this answers is not in the document]
Comment (vanderian) [reply testPad/compound.testpad.test.ts:9]: will move

File: testPad/compound.testpad.test.ts [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+
+  it("auto-compound: schedule & execute task", async () => {
+    testLog.getLog().info("running section: prepare compound extrinsic");
+
+    // @ts-ignore
+    const proxy = await oakApi.api.rpc.xcmpHandler.crossChainAccount(
+      user1.keyRingPair.address
+    );
+    const assetId = await getNextAssetId();
+    const poolId = assetId.add(BN_ONE);
```
[the remark this answers is not in the document]
Comment (vanderian) [reply testPad/compound.testpad.test.ts:122]: not a CI test, just a manual scenario

File: testPad/compound.testpad.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (vanderian) [reply testPad/compound.testpad.test.ts:247]: not a CI test, just a manual scenario

File: utils/eventListeners.ts [hunk 12]
```diff
@@ -127,3 +130,90 @@ export function findEventData(
...
+        if (counter === blocks) {
+          reject(`method ${method} not found within blocks limit`);
+        }
+      });
+    });
+  });
+};
+
+// used for APIs other than mangata
+export const signSendSuccess = async (
```
[the remark this answers is not in the document]
Comment (vanderian) [reply utils/eventListeners.ts:172]: yes

File: utils/Framework/Node/OakNode.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (vanderian) [reply utils/Framework/Node/OakNode.ts:4]: will fix

File: testPad/compound.testpad.test.ts [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+      ),
+      alice
+    );
+  }
+
+  it("auto-compound: schedule & execute task", async () => {
+    testLog.getLog().info("running section: prepare compound extrinsic");
+
+    // @ts-ignore
+    const proxy = await oakApi.api.rpc.xcmpHandler.crossChainAccount(
```
[the remark this answers is not in the document]
Comment (vanderian) [reply testPad/compound.testpad.test.ts:118]: we are using env var `const oakUri = process.env.OAK_URL;` to set the url to the API, you should set the
`export OAK_URL=....` to the node's url

File: utils/xToken.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  ChainId,
+  ChainSpecs,
+  TRANSFER_INSTRUCTIONS,
+  WEIGHT_IN_SECONDS,
+} from "./ChainSpecs";
+import { BN } from "@polkadot/util";
+import { User } from "./User";
+
+export class XToken {
+  static xcmTransferFee(toChain: ChainId, assetId: AssetSpec): BN {
```
Comment (goncer) [inline utils/xToken.ts:13]: Should we move `xTokenTransfer`  from OAK node to this class?

File: utils/xyk.ts [hunk 15]
```diff
@@ -57,7 +58,7 @@ export class Xyk {
     soldAssetId: BN,
     boughtAssetId: BN,
     boughtAssetAmount: BN,
-    maxAmountIn: BN = new BN(Number.MAX_SAFE_INTEGER)
+    maxAmountIn: BN = new BN("340282366920938463463374607431768211455") //u128::MAX
```
Comment (goncer) [inline utils/xyk.ts:61]: what is the difference? :O

File: utils/xToken.ts (hunk 14 above)
Comment (vanderian) [reply utils/xToken.ts:13]: naah, not worth it, would have to rector to have 'api' as argument as well
and kind a like it this way, separate the OAK impl just in case

File: utils/xyk.ts (hunk 15 above)
Comment (vanderian) [reply utils/xyk.ts:61]: max safe int:                                                                  9_007_199_254_740_991
u128::MAX.    340_282_366_920_938_463_463_374_607_431_768_211_455
the first one is less then 1MG with 18 decimals, not usable for our `Balance: u128` type

**Review by goncer — APPROVED** [review]
